addons/mass_mailing/static/src/js/mailing_m2o_filter.js

```diff
@@ -0,0 +1,158 @@
+import { Many2OneField } from '@web/views/fields/many2one/many2one_field';
+import { useService } from "@web/core/utils/hooks";
+import Domain from 'web.Domain';
+
+const { onMounted, onPatched, useExternalListener } = owl;
+
+/**
+ * Widget to create / remove favorite filters on mass mailing and/or marketing automation, extended
+ * from Many2OneField. This widget is designed specifically for 'mailing_filter_id'
+ * field on 'mailing.mailing' and 'marketing.campaign' form view.
+ *
+ * In edit mode, it will allow to save the latest configured domain
+ * in form of the favorite filter, or to remove the store filters.
+ *
+ */
+export class FieldMany2OneMailingFilter extends Many2OneField {
+    setup() {
+        super.setup();
+        this.notification = useService("notification");
+        onMounted(this._updateFilterIcons);
+        onPatched(this._updateFilterIcons);
+        useExternalListener(document, "show.bs.dropdown", this.onFilterDropdownShown);
+    }
+
+    /**
+     * Updates the 'Add to favorite' / 'Remove' icons' visibility based on the
+     * current state, and shows the custom message when no filter is available
+     * for the selected model.
+     *
+     * The filter can be saved if one of those conditions is matched:
+     * - No favorite filter is currently set
+     * - User emptied the input
+     * - User changed the domain when favorite filter is set
+     * - The input is currently being edited, known by the "this.state.isFloating" variable
+     *
+     * @private
+     */
+    _updateFilterIcons() {
+        const el = document.querySelector('.o_mass_mailing_filter_container');
+        if (!el || this.props.readonly) {
+            return;
+        }
+        const filterCount = this.props.record.data.mailing_filter_count;
+        el.querySelector('.o_mass_mailing_no_filter').classList.toggle('d-none', filterCount);
+        document.querySelectorAll('.o_field_many2one_selection > .o_input_dropdown')[1].classList.toggle('d-none', !filterCount);
+        // By default, domains in recordData are in string format, but adding / removing a leaf from domain widget converts
+        // value into object, so we use 'Domain' class to convert them in same (string) format, allowing proper comparison.
+        const recordDomain = new Domain(this.props.record.data[this.props.domain_field] || []).toString();
+        const filterDomain = new Domain(this.props.record.data.mailing_filter_domain || []).toString();
+        el.classList.toggle('d-none', recordDomain === '[]');
+        const canSaveFilter = !this.props.record.data.mailing_filter_id
+            || !document.querySelector(`input#${this.props.model_field}`).value.length
+            || this.state.isFloating
+            || filterDomain !== recordDomain;
+        el.querySelector('.o_mass_mailing_save_filter_container').classList.toggle('d-none', !canSaveFilter);
+        el.querySelector('.o_mass_mailing_remove_filter').classList.toggle('d-none', canSaveFilter);
```

Comment:
Wondering if we could use magic "state" for this.

Like having a t-attf-style in the DOM checking a this.state.canSaveFilter and this.state.filterCount that would be magically adapted.

Could be interesting to try it out to lessen manual DOM manipulations.